contents/android/IntroToAndroid.md

```diff
@@ -0,0 +1,108 @@
+
+The following code is an example of overriding the `onCreate` callback. It calls the superclass method first and then declares the user interface which is defined in an XML layout file `R.layout.activity_main`.
+
+```
+public class MainActivity extends Activity {
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+    }
+```
+
+### 2.4 MVVM and Architecture Components
+At Google I/O 2017, [Architecture Components](https://developer.android.com/topic/libraries/architecture/index.html) were introduced to help developers solve common issues by following best practices and implementing recommended architecture.
+
+#### LiveData
+Some examples of useful components for developers include the `LiveData` observable data holder class. Android apps usually use [RxJava](https://github.com/ReactiveX/RxJava/wiki/How-To-Use-RxJava) to do [reactive programming](https://blog.mindorks.com/rxjava-anatomy-what-is-rxjava-how-rxjava-is-designed-and-how-rxjava-works-d357b3aca586), which extends the `observable pattern` of Software Engineering. However, now with `LiveData`, we are able to create observables that are `lifecycle-aware`, only updating app components that are active. This prevents memory leaks and crashes. For more information, see the [LiveData page](https://developer.android.com/topic/libraries/architecture/livedata.html).
+
+#### ViewModel
+Another very useful component is the `ViewModel` component. It manages and store UI data in a `lifecycle-aware` way, allowing data to be retained during configuration changes such as screen rotations. Using normal UI controllers require data to be saved and then restored, but this is not practical for large amounts of data. For more information, see the [ViewModel page](https://developer.android.com/topic/libraries/architecture/viewmodel.html).
+
+#### How does this link to MVVM?
+Previously, Android apps were usually developed using the `Model-View-Presenter` (MVP) pattern, which is a derivative from the popular Model-View-Controller (MVC) pattern. However, this will change with the above `LiveData` and `ViewModel` components. Now, we can create apps that follow the `Model-View-ViewModel` (MVVM) architecture instead. The `ViewModel` component provided by Android serves as a good base for our ViewModel in the MVVM architecture. Our ViewModel in MVVM will interact with the Model and manages and store data to be observed by a View. The ViewModel is not aware about the View that is getting data from it to update the UI and thus it is decoupled from the View. This is very unlike the Presenter in the MVP pattern, which directly manipulates the View. To read more about the MVVM architecture in Android apps, see this [tutorial](https://proandroiddev.com/mvvm-architecture-viewmodel-and-livedata-part-1-604f50cda1).
```

Comment:
Actually, I think Android started with a modified version of MVC where everyone just floods activities or fragments with all sorts of logic. It's not exactly MVC as well, so maybe the distinction isn't that important.

Maybe it might be good to briefly mention the downfalls of MVC, MVP etc, or throw a link about it.